packages/keycloak-user-management/src/components/CreateEditUser/index.tsx

```diff
@@ -62,72 +64,97 @@ const CreateEditUser: React.FC<CreateEditPropTypes> = (props: CreateEditPropType
     practitioner: undefined,
   });
 
-  const { keycloakUser, keycloakBaseURL, opensrpBaseURL, extraData } = props;
+  const {
+    keycloakUser,
+    keycloakBaseURL,
+    opensrpBaseURL,
+    extraData,
+    fetchKeycloakUsersCreator,
+  } = props;
+
   const userId = props.match.params[ROUTE_PARAM_USER_ID];
 
   useEffect(() => {
-    if (keycloakUser && !initialValues.username) {
-      setInitialValues({ ...initialValues, ...keycloakUser });
+    if (keycloakUser) {
+      /** only update the object diff */
+      setInitialValues((prevState) => ({ ...prevState, ...keycloakUser }));
     }
-  }, [keycloakUser, initialValues]);
+  }, [keycloakUser]);
 
   useEffect(() => {
     if (!userGroups.length) {
+      setLoading(true);
       const serve = new KeycloakService(KEYCLOAK_URL_USER_GROUPS, keycloakBaseURL);
       serve
         .list()
         .then((response: UserGroup[]) => setUserGroups(response))
-        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED));
+        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED))
+        .finally(() => setLoading(false));
     }
   }, [keycloakBaseURL, opensrpBaseURL, userGroups.length]);
 
   /**
-   * Fetch user incase the user is not available e.g when page is refreshed
+   * Fetch user if userId changes (editing a different user)
    */
   useEffect(() => {
-    if (userId && !keycloakUser) {
+    if (userId) {
+      setLoading(true);
       const serve = new KeycloakService(KEYCLOAK_URL_USERS, keycloakBaseURL);
       serve
         .read(userId)
         .then((response: KeycloakUser | null | undefined) => {
-          if (response) fetchKeycloakUsers([response]);
+          if (response) fetchKeycloakUsersCreator([response]);
         })
-        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED));
+        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED))
+        .finally(() => setLoading(false));
     }
-  }, [userId, keycloakBaseURL, keycloakUser]);
+  }, [userId, keycloakBaseURL, fetchKeycloakUsersCreator]);
 
   /**
-   * Fetch User group of the user
+   * Fetch User group of the user being edited
    */
   useEffect(() => {
-    if (userId && initialValues.userGroup === undefined) {
+    if (userId) {
+      setLoading(true);
       const serve = new KeycloakService(
         KEYCLOAK_URL_USERS + '/' + userId + KEYCLOAK_URL_USER_GROUPS,
         keycloakBaseURL
       );
       serve
         .list()
         .then((response: UserGroup[]) =>
-          setInitialValues({ ...initialValues, userGroup: response.map((tag) => tag.id) })
+          setInitialValues((prevState) => ({
+            ...prevState,
+            userGroup: response.map((tag) => tag.id),
+          }))
         )
-        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED));
+        .catch((_: Error) => sendErrorNotification(lang.ERROR_OCCURED))
+        .finally(() => setLoading(false));
```

Comment:
@machariamuguku looks like we are calling setLoading asynchronously in more than one  place, implicetely introduces a race condition that where we could be setting an incomplete loading state, should we avoid this?

Reply:
The race condition is on setLoading which is overridden/skipped depending on if the other call is made with a similar or a different value. Assuming the first request sets Loading to true, and the second one attempts to set the same to true, the diffing algorithm takes charge and state is not updated (no state update, no rerender). Loading state is only updated if it's previous state is different than the one being set (in this case setLoading(true) before the first request and setLoading(false) after the last request)

Reply:
Alternatively we could do `if (!loading) setLoading(true)` before requests and `if (loading) setLoading(false)` before requests?

Reply:
Above confirmed by watching the loading state 
<img width="194" alt="Screenshot 2021-03-30 at 13 15 50" src="https://user-images.githubusercontent.com/10197807/112973728-4d9ab200-915a-11eb-907c-6a620fbdc9cd.png">

loading is only updated twice. First value is initial state, second is first update (to true), and third is last update (to false)
<img width="296" alt="Screenshot 2021-03-30 at 13 19 11" src="https://user-images.githubusercontent.com/10197807/112973961-8e92c680-915a-11eb-99ab-8207ca0f3b28.png">

Reply:
I was thinking more on the case where you set loading to false while some promise is actually loading, i.e. one promise resolves before the other.

Reply:
Any suggestions to avert that?